File: modules/cubex-scheduler/src/main/java/org/cubexmc/scheduler/ManagedCubexTask.java
package org.cubexmc.scheduler;

import com.tcoded.folialib.wrapper.task.WrappedTask;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicBoolean;
import org.bukkit.scheduler.BukkitTask;

final class ManagedCubexTask implements CubexTask {

    private final Runnable unregister;
    private final AtomicBoolean cancelled = new AtomicBoolean();
    private volatile Object nativeHandle;

    ManagedCubexTask(Runnable unregister) {
        this.unregister = unregister;
    }

    void attach(Object nativeHandle) {
        this.nativeHandle = nativeHandle;
        if (isCancelled()) {
            cancelNative(nativeHandle);
        }
    }

    void complete() {
        unregister.run();
    }

    @Override
    public void cancel() {
        if (cancelled.compareAndSet(false, true)) {
            cancelNative(nativeHandle);
            unregister.run();
        }
    }

    @Override
    public boolean isCancelled() {
        if (cancelled.get()) {
            return true;
        }
        Object handle = nativeHandle;
        if (handle instanceof WrappedTask) {
            return ((WrappedTask) handle).isCancelled();
        }
        if (handle instanceof BukkitTask) {
            return ((BukkitTask) handle).isCancelled();
        }
        if (handle instanceof CompletableFuture) {
            return ((CompletableFuture<?>) handle).isCancelled();
        }
        return false;
    }

    @Override
    public Object nativeHandle() {
        return nativeHandle;
    }

    static void cancelNative(Object handle) {
        if (handle == null) {
            return;
        }
        if (handle instanceof CubexTask) {
            ((CubexTask) handle).cancel();
            return;
        }
        if (handle instanceof WrappedTask) {
            ((WrappedTask) handle).cancel();
            return;
        }
        if (handle instanceof BukkitTask) {
            ((BukkitTask) handle).cancel();
            return;
        }
        if (handle instanceof CompletableFuture) {
            ((CompletableFuture<?>) handle).cancel(false);
            return;
        }
        try {
            handle.getClass().getMethod("cancel").invoke(handle);
        } catch (ReflectiveOperationException ignored) {
        }
    }
}

File: modules/cubex-scheduler/src/main/kotlin/org/cubexmc/scheduler/ManagedCubexTask.kt
package org.cubexmc.scheduler

import com.tcoded.folialib.wrapper.task.WrappedTask
import org.bukkit.scheduler.BukkitTask
import java.util.concurrent.CompletableFuture
import java.util.concurrent.atomic.AtomicBoolean

internal class ManagedCubexTask(private val unregister: Runnable) : CubexTask {
    private val cancelled = AtomicBoolean()

    @Volatile
    private var handle: Any? = null

    fun attach(nativeHandle: Any?) {
        handle = nativeHandle
        if (isCancelled()) cancelNative(nativeHandle)
    }

    fun complete() = unregister.run()

    override fun cancel() {
        if (cancelled.compareAndSet(false, true)) {
            cancelNative(handle)
            unregister.run()
        }
    }

    override fun isCancelled(): Boolean {
        if (cancelled.get()) return true
        return when (val nativeHandle = handle) {
            is WrappedTask -> nativeHandle.isCancelled
            is BukkitTask -> nativeHandle.isCancelled
            is CompletableFuture<*> -> nativeHandle.isCancelled
            else -> false
        }
    }

    override fun nativeHandle(): Any? = handle

    companion object {
        @JvmStatic
        fun cancelNative(handle: Any?) {
            when (handle) {
                null -> return
                is CubexTask -> handle.cancel()
                is WrappedTask -> handle.cancel()
                is BukkitTask -> handle.cancel()
                is CompletableFuture<*> -> handle.cancel(false)
                else -> try {
                    handle.javaClass.getMethod("cancel").invoke(handle)
                } catch (_: ReflectiveOperationException) {
                    // Unknown task handles are intentionally ignored by the compatibility adapter.
                }
            }
        }
    }
}
